### PR TITLE
Support ConnectionID Extension DTLS Negotiation

### PR DESCRIPTION
**Pull request by rizlik**
# Description
This PR adds initial partial support for ConnectionID as described by RFC9146 and RFC9147.
TLDR;
A pair of "labels" (the ConnectionIDs) is negotiated in the handshake and the labels will be attached to each record. Each peer negotiates the ConnectionID for the record it will receive (the RX path). In the future, ConnectionIDs may be used to decouple 
(src,dst,srcport,dstport) 4-tuple from the connection and to allow connections that survives to IP address changes (so it will support long-lasting connections behind a NAT).

Caveats of this PR:
- It adds support for DTLSv1.3 only
- It doesn't support RequestConnectionID/NewConnectionID messages
- No easy way to use the ConnectionIDs

# Testing

I added some tests to be sure the extension doesn't break connectivity.
I added a very basic test to check if the connectionID is negotiated correctly (`scripts/dtlscid.test`).
I **manually** with Wireshark that the ConnectionID is inserted in the record headers. An automatic test will be provided in the future.

# Checklist

 - [X] added tests
 - [X] updated/added doxygen

### CODE REVIEW

File: scripts/dtlscid.test [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+SERVER_PORT=11111
+
+(killall udp_proxy || true)
+(killall lt-server || true)
+(killall lt-client || true)
+
+test_cid () {
+    $WOLFSSL_ROOT/examples/server/server -v4 -u --cid $SCID 2>/tmp/serr 1>/tmp/sout &
+    SERVER_PID=$!
+    sleep 0.2
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply scripts/dtlscid.test:33]: This script does pretty much the same thing that `test-dtls13-cid.conf` does. I think it would be better to expand this script to test a multiplexing server (once implemented).

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
Overall the functionality looks great. My major concern is that this is unusable for servers. There is no way to multiplex packets on one socket based on the CID. As we discussed, it would be good to have a meta listener that keeps a `CID -> WOLFSSL` dictionary which can inject the DTLS message into the correct `WOLFSSL` object.

**Review by dgarske — CHANGES_REQUESTED** [review]
Great work!

File: examples/client/client.c [hunk 2]
```diff
@@ -3922,6 +3957,31 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
     }
 #endif
 
+#ifdef WOLFSSL_DTLS_CID
+    if (useDtlsCID && wolfSSL_dtls_cid_is_enabled(ssl)) {
```
[the remark this answers is not in the document]
Comment (rizlik) [reply examples/client/client.c:3961]: if the extension isn't negotiated, the `CIDInfo` structure is freed, so you can't check `CIDInfo->negotiated`.

File: scripts/dtlscid.test [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+trap cleanup err exit
+
+CCID="AA"
+SCID="BB"
+WOLFSSL_ROOT=$(pwd)
+SERVER_PORT=11111
+
+(killall udp_proxy || true)
+(killall lt-server || true)
+(killall lt-client || true)
```
[the remark this answers is not in the document]
Comment (rizlik) [reply scripts/dtlscid.test:28]: good point! I don't think so. But IMO it's better to address this, the fix to `dtls13.test` and the related port overlap issue in a separate PR

File: scripts/dtlscid.test (hunk 1 above)
[the remark this answers is not in the document]
Comment (rizlik) [reply scripts/dtlscid.test:33]: I agree with you, but I would do this in a separate PR as it's quite unrelated to CID and it will touch dtls13.test (and maybe tls13.test) as well.

File: src/dtls_cid.c [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    ret = TLSX_ConnectionID_Use(ssl);
+    if (ret != 0)
+        return ret;
+    return WOLFSSL_SUCCESS;
+}
+
+int wolfSSL_dtls_cid_is_enabled(WOLFSSL* ssl)
+{
+    return TLSX_Find(ssl->extensions, TLSX_CONNECTION_ID) != NULL;
+}
```
[the remark this answers is not in the document]
Comment (rizlik) [reply src/dtls_cid.c:268]: Now the `CIDInfo` pointer is stored in the SSL object itself

File: examples/client/client.c [hunk 5]
```diff
@@ -119,6 +119,10 @@ static const char kHttpGetMsg[] = "GET /index.html HTTP/1.0\r\n\r\n";
     #define TEST_DELAY() XSLEEP_MS(1000)
 #endif
 
+#ifdef WOLFSSL_DTLS_CID
+#define DTLS_CID_MAX_SIZE 256
```
[the remark this answers is not in the document]
Comment (rizlik) [reply examples/client/client.c:123]: I don't think it should be overridable, it's just needed to have a size for the buffer that will store the connection ID when parsing the options inside the examples. I renamed it to `DTLS_CID_BUFFER_SIZE`, moved it to `test.h` and unwrapped it from `WOLFSSL_DTLS_CID`

File: src/dtls13.c [hunk 6]
```diff
@@ -1237,23 +1313,24 @@ int Dtls13ParseUnifiedRecordLayer(WOLFSSL* ssl, const byte* input,
...
         return BAD_FUNC_ARG;
 
-    if (*input & DTLS13_CID_BIT) {
-        WOLFSSL_MSG("DTLS1.3 header with connection ID. Not supported");
-        return WOLFSSL_NOT_IMPLEMENTED;
-    }
-
+    flags = *input;
     idx = DTLS13_HDR_FLAGS_SIZE;
+    ret = Dtls13UnifiedHeaderParseCID(ssl, flags, input, inputSize, &idx);
```
[the remark this answers is not in the document]
Comment (rizlik) [reply src/dtls13.c:1325]: `DTLS13_HDR_FLAGS_SIZE` is 1 byte, so the check was implicitly in the `if (input == NULL || inputSize == 0)` check. I modified the check to make it more explicit.

File: src/internal.c [hunk 7]
```diff
@@ -9787,7 +9788,7 @@ static int GetDtls13RecordHeader(WOLFSSL* ssl, const byte* input,
             return SEQUENCE_ERROR;
     }
 
-    ret = Dtls13GetUnifiedHeaderSize(
+    ret = Dtls13GetUnifiedHeaderSize(ssl,
```
[the remark this answers is not in the document]
Comment (rizlik) [reply src/internal.c:9791]: it should be covered by:
```
    readSize = ssl->buffers.inputBuffer.length - *inOutIdx;

    if (readSize < DTLS_UNIFIED_HEADER_MIN_SZ)
        return BUFFER_ERROR;
```
at the beginning of the function.`DTLS_UNIFIED_HEADER_MIN_SZ` is 2.

**Comment by rizlik** [comment]
Thanks for the feedback. I addressed the comments and rebased.

File: src/dtls_cid.c [hunk 8]
```diff
@@ -0,0 +1,361 @@
+/* dtls_cid.c
```
Comment (dgarske) [inline src/dtls_cid.c:1]: Let's rename this to `dtls.c` and use it to put all DTLS specific functions. You will also need to add it to CMakeList.txt and the visual studio projects.

**Review by dgarske — CHANGES_REQUESTED** [review]
```
./configure --enable-dtls --enable-dtls13 --enable-dtlscid CC=g++ && make
...
src/dtls_cid.c: In function ‘int TLSX_ConnectionID_Parse(WOLFSSL*, const byte*, word16, byte)’:
./wolfssl/wolfcrypt/types.h:505:67: error: invalid conversion from ‘void*’ to ‘ConnectionID*’ [-fpermissive]
  505 |                 #define XMALLOC(s, h, t)     ((void)(h), (void)(t), wolfSSL_Malloc((s)))
      |                                              ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                                                   |
      |                                                                   void*
src/dtls_cid.c:248:14: note: in expansion of macro ‘XMALLOC’
  248 |         id = XMALLOC(sizeof(*id) + cidSize, ssl->heap, DYNAMIC_TYPE_TLSX);
      |              ^~~~~~~
make[2]: *** [Makefile:6263: src/libwolfssl_la-dtls_cid.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
src/tls.c: In function ‘void TLSX_FreeAll(TLSX*, void*)’:
src/tls.c:10654:33: error: invalid conversion from ‘void*’ to ‘byte*’ {aka ‘unsigned char*’} [-fpermissive]
10654 |             CID_FREE(extension->data, heap);
      |                      ~~~~~~~~~~~^~~~
      |                                 |
      |                                 void*
In file included from src/tls.c:33:
./wolfssl/internal.h:2894:49: note:   initializing argument 1 of ‘void TLSX_ConnectionID_Free(byte*, void*)’
 2894 | WOLFSSL_LOCAL void TLSX_ConnectionID_Free(byte* ext, void* heap);
      |                                           ~~~~~~^~~
src/tls.c: In function ‘int TLSX_GetSize(TLSX*, byte*, byte, word16*)’:
src/tls.c:10823:51: error: invalid conversion from ‘void*’ to ‘byte*’ {aka ‘unsigned char*’} [-fpermissive]
10823 |                 length += CID_GET_SIZE(extension->data);
      |                                        ~~~~~~~~~~~^~~~
      |                                                   |
      |                                                   void*
In file included from src/tls.c:33:
./wolfssl/internal.h:2896:54: note:   initializing argument 1 of ‘word16 TLSX_ConnectionID_GetSize(byte*)’
 2896 | WOLFSSL_LOCAL word16 TLSX_ConnectionID_GetSize(byte* ext);
      |                                                ~~~~~~^~~
src/tls.c: In function ‘int TLSX_Write(TLSX*, byte*, byte*, byte, word16*)’:
src/tls.c:11025:48: error: invalid conversion from ‘void*’ to ‘byte*’ {aka ‘unsigned char*’} [-fpermissive]
11025 |                 offset += CID_WRITE(extension->data, output+offset);
      |                                     ~~~~~~~~~~~^~~~
      |                                                |
      |                                                void*
In file included from src/tls.c:33:
./wolfssl/internal.h:2895:52: note:   initializing argument 1 of ‘word16 TLSX_ConnectionID_Write(byte*, byte*)’
 2895 | WOLFSSL_LOCAL word16 TLSX_ConnectionID_Write(byte* ext, byte* output);
      |                                              ~~~~~~^~~
src/tls13.c: In function ‘int SendTls13ServerHello(WOLFSSL*, byte)’:
src/tls13.c:5883:36: error: invalid conversion from ‘byte’ {aka ‘unsigned char’} to ‘HandShakeType’ [-fpermissive]
 5883 |                                    extMsgType, 0);
      |                                    ^~~~~~~~~~
      |                                    |
      |                                    byte {aka unsigned char}
In file included from src/tls13.c:104:
./wolfssl/internal.h:5666:59: note:   initializing argument 5 of ‘int Dtls13HandshakeSend(WOLFSSL*, byte*, word16, word16, HandShakeType, int)’
 5666 |     word16 output_size, word16 length, enum HandShakeType handshake_type,
      |                                        ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
```

File: tests/test-dtls13-cid.conf [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+# server DTLSv1.3 No CID support
+-v 4
+-u
+-l TLS_AES_128_GCM_SHA256
+
+# client DTLSv1.3 B CID
+-v 4
+-u
+-l TLS_AES_128_GCM_SHA256
+--cid B
```
Comment (julek-wolfssl) [inline tests/test-dtls13-cid.conf:58]: Is there a maximum CID length? Maybe let's test longer CID as well.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]

File: tests/test-dtls13-cid.conf (hunk 9 above)
Comment (rizlik) [reply tests/test-dtls13-cid.conf:58]: For now, the maximum CID length is a compile-time define, and it's on purpose quite small (2 bytes).
Unfortunately, at the moment we need to copy the received record header, and we are copying it in a static buffer. This means that the max length of the CID increases the per-session memory footprint. I think that a nice to have optimization would be to  avoid the copy and the static buffer altogether. After that, I think we can use a longer max size. For now, tests will fail with  CID longer than 2 bytes. I can increase the length to two bytes

Comment (rizlik) [reply tests/test-dtls13-cid.conf]: ```    byte dtls13CurRL[DTLS_RECVD_RL_HEADER_MAX_SZ];``` inside WOLFSSL object.

Comment (rizlik) [reply tests/test-dtls13-cid.conf]: Allocating the buffer will solve the memory problem, but we still need to make a copy, which can be inefficient. I think that a better, zero-copy solution, is to just refer to the current RL as it is inside the input buffer. Right now, it can't be done because we may read and discard the record (advance the input buffer index) before processing the full message (it can happen only if doing DTLS over a stream-oriented medium (TCP)). I think this is material for another PR btw.

Also, consider that this limitation is just for the ConnectionID on the RX path (the other peer can use up to 2**8-1 always). This means that there are several practical applications that can use a small value for the connection ID max size (a client can even use 0 if the IP of the server never changes but it wants to support the extension because it is behind a NAT).

**Review by julek-wolfssl — APPROVED** [review]
I like this as an initial solution. The major missing feature here is a way for a user server to multiplex on the CID.

File: scripts/dtlscid.test [hunk 10]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
Comment (dgarske) [inline scripts/dtlscid.test:1]: On Mac see a warning, but test still works:

```
% ./scripts/dtlscid.test
No matching processes belonging to you were found
No matching processes belonging to you were found
(standard_in) 1: illegal character: N
(standard_in) 1: illegal character: N
Sending CID is 4242
Sending CID is 4141
Cleaning up...
```

Please followup later with adding tests to the CI.

**Review by dgarske — APPROVED** [review]
Fixed a dynamic type typo for you. Please followup with making sure the DTLS CID tests are in PRB CI. Thanks
